### PR TITLE
Add mt8195 platform specific changes

### PR DESCRIPTION
**Pull request by kuanhsuncheng**
Hi sirs,

This is Allen Cheng from MediaTek.

There are some commits I don't put in the previous pull request.

Most of them will modify the common code in sof.

Some are hw restriction (compile error) and feature(mpu) when we develop mt8195 in chromebook.

If there is any comments, please let me know.

Thanks.

### CODE REVIEW

File: src/arch/xtensa/include/arch/atomic.h [hunk 1]
```diff
@@ -11,6 +11,9 @@
 #define __ARCH_ATOMIC_H__
 
 #include <stdint.h>
+#ifdef XCHAL_HAVE_EXCLUSIVE
+#include <xtensa/tie/xt_core.h>
```
Comment (lgirdwood) [inline src/arch/xtensa/include/arch/atomic.h:15]: So these are not available with GCC build. 
```
                 from /home/sof/work/sof.git/src/platform/tigerlake/lib/clk.c:9:
/home/sof/work/sof.git/src/arch/xtensa/include/arch/atomic.h:15:10: fatal error: xtensa/tie/xt_core.h: No such file or directory
   15 | #include <xtensa/tie/xt_core.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~
```
Can we make this a Kconfig option i.e. 

```#if XCHAL_HAVE_EXCLUSIVE && CONFIG_XTENSA_EXCLUSIVE && __XCC__``` means we only build this code if it's selected in Kconfig and we are using Cadence compiler (GCC would still build).

File: src/arch/xtensa/include/arch/atomic.h [hunk 2]
```diff
@@ -44,12 +48,28 @@ static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
...
+
+	while (!result) {
+		current = XT_L32EX((int32_t *)a);
+		result = current + value;
+		XT_S32EX(result, (int32_t *)a);
+		XT_GETEX(result);
+	}
+	return current;
+#else
+	// Not support atomic
```
Comment (lgirdwood) [inline src/arch/xtensa/include/arch/atomic.h:65]: This is not a problem on unicore UP systems, we  can just do 
```c
#elif CONFIG_NUM_CORES == 1
/* This is fine on UP systems */
*a += value;
#else 
#error No atomic methods available.
#endif
```
same for the others.

**Review by lgirdwood — COMMENTED** [review]
I think we just need the Kconfig selection of exclusive as this looks like its showing lots of red in CI (probably because it's not cache coherent like S32CLI on Intel DSPs).

File: src/arch/xtensa/include/arch/atomic.h [hunk 3]
```diff
@@ -63,6 +86,25 @@ static inline int32_t arch_atomic_sub(atomic_t *a, int32_t value)
...
+		XT_S32EX(result, (int *)a);
+		XT_GETEX(result);
+	}
+	return current;
+#elif CONFIG_NUM_CORES == 1
+	/* This is fine on unicore  UP systems */
+	*a += value;
+	return *a;
+#else
+	#error No atomic methods available.
```
Comment (kuanhsuncheng) [inline src/arch/xtensa/include/arch/atomic.h:106]: Hi @lgirdwood ,

Do I need to add XCHAL_HAVE_S32C1I?

if we use gcc to compile, it seems compiler will go to "#error No atomic methods available".

**Review by kuanhsuncheng — COMMENTED** [review]
Hi I add CONFIG_XTENSA_EXCLUSIVE  in kconfig,

let me know if there is anything needs to be changed, thanks.

File: src/arch/xtensa/include/arch/atomic.h (hunk 3 above)
Comment (lgirdwood) [reply src/arch/xtensa/include/arch/atomic.h:106]: Yes. we need 
```c
#if CONFIG_LOCK_EXCLUSIVE && __XCC__ && XCHAL_EXCLUSIVE
 /* do this */
#elif XCHAL_S32C1I
/* do that */
#elif CONFIG_NUM_CORES == 1
/* do something else */
#else 
#error No atomic support for multicore system
#endif
```

Comment (lgirdwood) [reply src/arch/xtensa/include/arch/atomic.h:106]: See the CI build
```
warning: the int symbol CORE_COUNT (defined at src/platform/Kconfig:272) has a non-int default MP_NUM_CPUS (undefined)
-- MEU_OFFSET=1344
-- Configuring done
-- Generating done
-- Build files have been written to: /srv/home/jenkins/workspace/sof_generic_build/build_tgl_xcc
[  0%] Built target genconfig
--
                 from /srv/home/jenkins/workspace/sof_generic_build/src/include/sof/lib/dma.h:20,
                 from /srv/home/jenkins/workspace/sof_generic_build/src/platform/intel/cavs/include/cavs/drivers/dw-dma.h:14,
                 from /srv/home/jenkins/workspace/sof_generic_build/src/platform/tigerlake/include/platform/drivers/dw-dma.h:13,
                 from /srv/home/jenkins/workspace/sof_generic_build/src/include/sof/drivers/dw-dma.h:12,
                 from /srv/home/jenkins/workspace/sof_generic_build/src/platform/intel/cavs/lib/dma.c:11:
/srv/home/jenkins/workspace/sof_generic_build/src/arch/xtensa/include/arch/atomic.h:69:3: error: #error No atomic methods available.
/srv/home/jenkins/workspace/sof_generic_build/src/arch/xtensa/include/arch/atomic.h:106:3: error: #error No atomic methods available.
[ 45%] Building C object CMakeFiles/sof.dir/src/platform/intel/cavs/lib/mem_window.c.o
[ 45%] Building C object CMakeFiles/sof.dir/src/platform/intel/cavs/platform.c.o
make[3]: *** [CMakeFiles/sof.dir/build.make:118: CMakeFiles/sof.dir/src/platform/intel/cavs/lib/dma.c.o] Error 2
```

**Comment by lgirdwood** [comment]
@andyross fyi - for Zephyr support of xtensa EXCLUSIVE locking and MPU.

File: src/arch/xtensa/include/arch/atomic.h (hunk 3 above)
Comment (kuanhsuncheng) [reply src/arch/xtensa/include/arch/atomic.h:118]: Hi Liam,

For gcc-build, there is no XCHAL_HAVE_EXCLUSIVE  and XCHAL_HAVE_S32C1I used, 

I wonder how it handles those cases before.

**Review by kuanhsuncheng — COMMENTED** [review]
HI I found there is no CONFIG_NUM_CORES macro define.

so I use CONFIG_CORE_COUNT to check the current core number.

let me know if there is anything needs to be changed, thanks.

File: src/arch/xtensa/include/arch/atomic.h (hunk 3 above)
Comment (lgirdwood) [reply src/arch/xtensa/include/arch/atomic.h:118]: That's a good point - then I think its best if we have for atomic.h and spinlock.h a set of functions for EXCLUSIVE and the existing set of functions for everything else. e.g.

```c
#if XCHAL_HAVE_EXCLUSIVE && CONFIG_XTENSA_EXCLUSIVE && __XCC__

/* All EXCLUSIVE methods here */

static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
{
	/*reference xtos : xipc_misc.h*/
	int32_t result = 0;
	int32_t current;

	while (!result) {
		current = XT_L32EX((int32_t *)a);
		result = current + value;
		XT_S32EX(result, (int32_t *)a);
		XT_GETEX(result);
	}
	return current;
}

#elif XCHAL_S32C1I

/* all S32C1I methods here */

static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
{
	int32_t result, current;

	__asm__ __volatile__(
		"1:     l32i    %1, %2, 0\n"
		"       wsr     %1, scompare1\n"
		"       add     %0, %1, %3\n"
		"       s32c1i  %0, %2, 0\n"
		"       bne     %0, %1, 1b\n"
		: "=&a" (result), "=&a" (current)
		: "a" (&a->value), "a" (value)
		: "memory");

	return current;
}

#else 

/* The xtensa config has no atomic method ISA support - we can use normal integer arithmetic iff UP */
#if CONFIG_NUM_CORES > 1
#error No atomic ISA for SMP configuration
#endif

/* integer arithmetic methods here */

#endif
```
This is simple do to in both files and wont break anything.

File: src/arch/xtensa/exc-dump.S [hunk 4]
```diff
@@ -100,8 +100,10 @@ dump_special_registers:
 	s32i a6, a2, REG_OFFSET_EPC3
 	rsr  a6, EPC4
```
Comment (dbaluta) [inline src/arch/xtensa/exc-dump.S:101]: Can please put a snippet of the error message in the commit ?

File: src/arch/xtensa/xtos/reset-vector.S [hunk 5]
```diff
@@ -37,6 +37,10 @@
 #include <xtensa/xtruntime-core-state.h>
 #include "xtos-internal.h"
 
+#if XCHAL_HAVE_MPU
```
Comment (dbaluta) [inline src/arch/xtensa/xtos/reset-vector.S:40]: Please explain WHY this is needed.

File: scripts/xtensa-build-all.sh [hunk 6]
```diff
@@ -329,6 +329,12 @@ do
 			HOST="xtensa-rn-elf"
 			XTENSA_TOOLS_VERSION="RF-2016.4-linux"
 			;;
+		mt8195)
+			PLATFORM="mt8195"
+			XTENSA_CORE="hifi4_8195_PROD"
+			HOST="xtensa-mt8195-elf"
+			XTENSA_TOOLS_VERSION="RI-2019.1-linux"
+			;;
```
Comment (dbaluta) [inline scripts/xtensa-build-all.sh:338]: Modifying xtensa-build-all.sh should be a separate patch with a commit message like this

```
scripts: Add mt8195 to build script
```

File: src/drivers/mediatek/mt8195/CMakeLists.txt [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: BSD-3-Clause
 
 add_local_sources(sof ipc.c timer.c
-	interrupt.c uart.c afe-memif.c afe-dai.c afe-drv.c
+	interrupt.c afe-memif.c afe-dai.c afe-drv.c
 )
```
Comment (dbaluta) [inline src/drivers/mediatek/mt8195/CMakeLists.txt:6]: So, this was added by mistake. Please create a separate patch which removes the files added to build system by mistake.

**Review by kuanhsuncheng — COMMENTED** [review]
Update the new changes, 

If there is any comments, please let me know.

Thanks.

**Comment by dbaluta** [comment]
We can immediately merge the following patches:

- drivers: mtk: remove superfluous variable for mt8195
- build: remove unused files from build system
- Kconfig: add mt8195 to Kconfig 
- scripts: Add mt8195 to build script 

if you move them to another PR and leave time for others to review the rest of the patches which look to be a little bit more delicate :)

File: src/arch/xtensa/include/arch/atomic.h [hunk 8]
```diff
@@ -65,6 +105,39 @@ static inline int32_t arch_atomic_sub(atomic_t *a, int32_t value)
...
+/* integer arithmetic methods */
+static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
+{
+	int32_t result;
+	int32_t current;
+
+	current = arch_atomic_read(a);
+	result = current - value;
+	arch_atomic_set(a, value);
+	return result;
```
Comment (lgirdwood) [inline src/arch/xtensa/include/arch/atomic.h:125]: This should return current value, this way the caller know the value at the addition and the result (in a*).

File: src/arch/xtensa/include/arch/atomic.h [hunk 9]
```diff
@@ -65,6 +105,39 @@ static inline int32_t arch_atomic_sub(atomic_t *a, int32_t value)
...
+
+static inline int32_t arch_atomic_sub(atomic_t *a, int32_t value)
+{
+	int32_t result;
+	int32_t current;
+
+	current = arch_atomic_read(a);
+	result = current - value;
+	arch_atomic_set(a, value);
+	return result;
```
Comment (lgirdwood) [inline src/arch/xtensa/include/arch/atomic.h:136]: Should return current.

**Review by lgirdwood — COMMENTED** [review]
@kuanhsuncheng these are all good to merge except the atomic one as it needs an update.

**Comment by kuanhsuncheng** [comment]
> We can immediately merge the following patches:
> 
> * drivers: mtk: remove superfluous variable for mt8195
> * build: remove unused files from build system
> * Kconfig: add mt8195 to Kconfig
> * scripts: Add mt8195 to build script
> 
> if you move them to another PR and leave time for others to review the rest of the patches which look to be a little bit more delicate :)

hi Daniel,

I agree that, it will be more clear for PR.

I create another PR for MT8196 build of sof.

https://github.com/thesofproject/sof/pull/4741

Thanks for your comment.

**Review by kuanhsuncheng — COMMENTED** [review]
Hi 

I update the new version.

Add spin lock simple implement for for UP

Also move build script patch to other PR.

let me know if there is anything needs to be changed, thanks.

**Comment by lgirdwood** [comment]
CI showing unrelated timeouts on alsabat and suspend unrelated to this PR.

**Review by lgirdwood — APPROVED** [review]

File: src/arch/xtensa/include/arch/spinlock.h [hunk 10]
```diff
@@ -72,6 +152,28 @@ static inline void arch_spin_unlock(spinlock_t *lock)
...
+/*
+ * The ISA has no atomic operations so use integer arithmetic on uniprocessor systems.
+ * This helps support GCC and qemu emulation of certain targets.
+ */
+static inline void arch_spin_unlock(spinlock_t *lock)
+{
+	uint32_t result;
+
+	lock->lock = 0;
+	result = 1;
```
Comment (marc-hb) [inline src/arch/xtensa/include/arch/spinlock.h:172]: Unused `result` variable, removal submitted in #9351
